spec/javascript/packages/document-capture/components/document-capture-warning-spec.jsx

```diff
@@ -0,0 +1,595 @@
+
+        expect(getByText(/general error/)).to.be.ok();
+        expect(getByText(/idv.warning.attempts_html/)).to.be.ok();
+        expect(getByRole('button', { name: 'idv.failure.button.try_online' })).to.be.ok();
+
+        // ipp section
+        expect(
+          screen.getByRole('heading', {
+            name: 'in_person_proofing.headings.cta',
+            level: 2,
+          }),
+        ).to.be.ok();
+        /* ipp prompt */
+        expect(getByText('in_person_proofing.body.cta.prompt_detail')).to.be.ok();
+        expect(getByRole('button', { name: 'in_person_proofing.body.cta.button' })).to.be.ok();
+
+        // troubleshooting section
+        expect(
+          screen.getByRole('heading', {
+            name: 'components.troubleshooting_options.ipp_heading',
+            level: 2,
+          }),
+        ).to.be.ok();
+        /* list of troubleshooting links */
+        const troubleShootingList = getByRole('list');
+        expect(troubleShootingList).to.be.ok();
+
+        const troubleShootingListItem = within(troubleShootingList).getAllByRole('listitem');
+        expect(troubleShootingListItem).to.be.ok();
+        const hrefLinks = within(troubleShootingList).getAllByRole('link');
+
+        hrefLinks.forEach((hrefLink) => {
+          expect(hrefLink.href).to.contains('location=post_submission_warning');
+          expect(hrefLink.href).to.contains('article=');
+        });
+      });
+    });
+
+    //
+    context('failed result', () => {
+      const isFailedResult = true;
+      it('renders not failed doc type', () => {
+        const isFailedDocType = false;
+        const unknownFieldErrors = [
+          {
+            field: 'general',
+            error: toFormEntryError({ field: 'general', message: 'general error' }),
+          },
+        ];
+        const { getByRole, getByText } = render(
+          <InPersonContext.Provider value={{ inPersonURL: '/verify/doc_capture' }}>
+            <DocumentCaptureWarning
+              isFailedDocType={isFailedDocType}
+              isFailedResult={isFailedResult}
+              remainingAttempts={2}
+              unknownFieldErrors={unknownFieldErrors}
+              actionOnClick={() => {}}
+            />
+            ,
+          </InPersonContext.Provider>,
+        );
+
+        // error message section
+        const h1 = screen.getByRole('heading', {
+          name: 'errors.doc_auth.rate_limited_heading',
+          level: 1,
+        });
+        expect(h1).to.be.ok();
+
+        const h2 = screen.queryByRole('heading', {
+          name: 'errors.doc_auth.rate_limited_subheading',
+          level: 2,
+        });
+        expect(h2).to.be.null();
+
+        expect(getByText('general error')).to.be.ok();
+        expect(getByText('idv.failure.attempts_html')).to.be.ok();
+        expect(getByRole('button', { name: 'idv.failure.button.warning' })).to.be.ok();
+
+        // the ipp section is not displayed with isFailedResult=true
```

Comment:
I do wonder if maybe some of these should be separate `it` blocks, as they seem pretty long for one test (and comments like this in particular feel like an `it` block).

But, I am less familiar with React and testing in React so maybe the length is necessary. It might be good to have someone more familiar with React to review the tests.

Reply:
Couple ideas for shortening / making reusable:

- Where we're not expecting content to exist, one assertion should suffice, rather than checking the heading, prompt, and button? i.e. if they're all statically rendered via InPersonCallToAction , checking one should be enough.
   - It's not really well supported in our toolchain, but this is where something like shallow-rendering and checking for the rendered existence of some child component would be helpful (e.g. [Enzyme shallow rendering with Component finder](https://enzymejs.github.io/enzyme/docs/api/ShallowWrapper/find.html#examples))
- Create functions to either assert, render common elements, or add common test cases
    - Not React-specific, but some generic example patterns to consider following: [`itAcceptsAValidDate`](https://github.com/18F/identity-idp/blob/b6ca71464af7f2134c679c571ea5e855f5595e2a/app/javascript/packages/memorable-date/index.spec.ts#L153-L163), [`createElement`](https://github.com/18F/identity-idp/blob/b6ca71464af7f2134c679c571ea5e855f5595e2a/app/javascript/packages/countdown/countdown-alert-element.spec.ts#L8)
- Have a single spec test the exhaustive set of content expected, then each variation only needs to a small part of that, if it's expected that the content would come together
   - e.g. rather than repeat [detailed assertions for troubleshooting links](https://github.com/18F/identity-idp/blob/597986c2788f17efd70786502bc240d5e57362ee/spec/javascript/packages/document-capture/components/document-capture-warning-spec.jsx#L283-L294), do the detailed assertion once, and then everywhere else can just check for a list?
      - Or put those assertions into a function?

Reply:
@aduth good suggestion.

Reply:
@aduth  refactored the test with reusable functions.